packages/react-components/src/components/xy-output-component.tsx

```diff
@@ -426,24 +432,93 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
         }
     }
 
+    private tooltip(x: number, y: number): void {
+        const xPos = this.positionXMove;
+        const timeForX = this.getTimeForX(xPos);
+        let timeLabel: string | undefined = timeForX.toString();
+        if (this.props.unitController.numberTranslator) {
+            timeLabel = this.props.unitController.numberTranslator(timeForX);
+        }
+        const chartWidth = this.lineChartRef.current.chartInstance.width;
+        const arraySize = this.state.xyData.labels.length;
+        const index = Math.max(Math.round((xPos / chartWidth) * (arraySize - 1)), 0);
+        const points: any = [];
+        let zeros = 0;
+        this.state.xyData.datasets.forEach((d: any) => {
+            // In case there are less data points than pixels in the chart,
+            // calculate nearest value.
+            const adjustedY = d.data[index];
+            const rounded = isNaN(adjustedY) ? 0 : (BIMath.round(Number(adjustedY) * 100)/BIMath.round(100));
+            if (this.state.xyData.datasets.length === 1 || rounded > 0) {
+                const point: any = {
+                    label: d.label,
+                    color: d.borderColor,
+                    background: d.backgroundColor,
+                    value: rounded.toString(),
+                };
+                points.push(point);
+            }
+            else {
+                zeros += 1;
+            }
+        });
+        // Sort in decreasing order
+        points.sort((a: any, b: any) => Number(b.value) - Number(a.value));
+
+        // Adjust tooltip position if mouse is close to the bottom of the window
+        let topPos = undefined;
+        let bottomPos = undefined;
+        if (y > window.innerHeight - 270) {
```

Comment:
This only works if the tooltip height is guaranteed to be less than 270, is it the case?

Otherwise maybe safer to use innerheight / 2 as a condition? Although, that wastes half the space in every case. So it wouldn't work if the available height is less than twice the tooltip size. That is worst. never mind...

I don't know if there's a way to know the exact height of the tooltip before computing the position? Otherwise calculate an approximation based on the number of rows of text it will contain?

In any case, I would also suggest to invert logic so that if there is not enough height, topPos is used instead of bottomPos, as I think it's better to clip the bottom than to clip the top?

Reply:
We have an upper limit to the size of the tooltip (`max-height = 250px` in tooltip-xy-component), so using 250 here is safe and more precise than inneheight / 2.

Reply:
I can reproduce with the maximum number of items in the tooltip and just at the threshold of top/bottom condition a tooltip below the pointer that is clipped for last 1 or 2 text lines.

Reply:
Increased the threshold. Now the tooltip is never clipped.